### PR TITLE
Feature: Supporting Map, Tuple, List.

### PR DESCRIPTION
**Pull request by AbnerZheng**
### Overview

This PR add supports for `Map`, `Tuple`, and `List`(I call them complex types) in Clarity.  Currently, it doesn't support nest complex type, for example, Map<`Tuple`, `List`>,  but we support `bytes` as elements of these three types, even though `bytes` are `List<u8>` in CLType. It is because we could represent `byte` in base16 encoded string. 

This PR also include a lot of unit test for `Arg` DSL, and `DeployArgumentParser`.  The `Arg` DSL enables us to create CLType, CLValue, and CLInstance very fast, I includes this DSL to `casperlab-sdk`. And inside DeployArgumentParser, there are two kinds of static methods, one is validating methods, another is building methods. When users try to create arguments, he needs input value in JSON, and then we would validate the value basing the types he select and if there is no problem, we would call building methods to construct `CLValueInstance`.

### Which JIRA ticket does this PR relate to?
https://casperlabs.atlassian.net/browse/ECO-513
https://casperlabs.atlassian.net/browse/ECO-611
https://casperlabs.atlassian.net/browse/ECO-612

### Complete this checklist before you submit this PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [X] This PR meets [CasperLabs coding standards](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/16842753/Coding+Standards).
- [X] If this PR adds a new feature, it includes tests related to this feature.
- [X] You assigned one person to review this PR.
- [X] Your GitHub account is linked with our [Drone CI](https://drone-auto.casperlabs.io/) system. This is necessary to run tests on this PR.
- [X] Do not forget to run `bors r+` if GitHub policy is not enforced, e.g. when merging into another feature branch. It may be omitted under some circumstances if this PR intentionally assumes that integration tests will fail but will be fixed with the future PRs.

### Notes
_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._

### CODE REVIEW

File: packages/ui/src/lib/DeployArgumentParser.ts [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+      case 'Bytes':
+        return DeployArgumentParser.validateBase16String(argValueInJson);
+      case 'Bytes (Fixed Length)':
+        return DeployArgumentParser.validateBase16String(argValueInJson);
+    }
+    return false;
+  }
+
+  static validateBase16String(v: any): string | false {
+    const valid = typeof v === typeof 's';
```
Comment (zie1ony) [inline packages/ui/src/lib/DeployArgumentParser.ts:231]: Shouldn't that be a `typeof v === 'string'` ?

File: packages/ui/src/lib/DeployArgumentParser.ts [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+        );
+        break;
+      case 'FixedList':
+        clValueInstance = this.buildListTypeArg(
+          arg.$.listInnerDeployArgs.$,
+          argValueStrInJson,
+          true
+        );
+        break;
+      case 0:
```
Comment (zie1ony) [inline packages/ui/src/lib/DeployArgumentParser.ts:455]: What those numbers stand for?

**Review by zie1ony — APPROVED** [review]
Great job Abner. Nothing I can complain about. Just two questions.
Sadly you'll have to rewrite all of that to be usable with node-rs.

File: packages/ui/src/lib/DeployArgumentParser.ts (hunk 1 above)
Comment (AbnerZheng) [reply packages/ui/src/lib/DeployArgumentParser.ts:231]: yes, they are the same.  type of 's' return 'string.

File: packages/ui/src/lib/DeployArgumentParser.ts (hunk 2 above)
Comment (AbnerZheng) [reply packages/ui/src/lib/DeployArgumentParser.ts:455]: They are from `CLType.Simple`, represent simple types like u8, u16, string and so on.

**Pull request merged**